File: python-ml-service/feature_engineering.py
"""
Professional ML Feature Engineering for Stock/ETF Prediction
Extracts 40+ features from historical price data
"""

import pandas as pd
import numpy as np
from typing import Dict, List, Tuple


def calculate_rsi(prices: pd.Series, period: int = 14) -> float:
    """Calculate RSI using Wilder's smoothing method"""
    delta = prices.diff()
    gain = (delta.where(delta > 0, 0)).rolling(window=period).mean()
    loss = (-delta.where(delta < 0, 0)).rolling(window=period).mean()

    rs = gain / loss
    rsi = 100 - (100 / (1 + rs))
    return rsi.iloc[-1] if not pd.isna(rsi.iloc[-1]) else 50.0


def calculate_macd(prices: pd.Series, fast: int = 12, slow: int = 26, signal: int = 9) -> Dict[str, float]:
    """Calculate MACD indicator"""
    ema_fast = prices.ewm(span=fast, adjust=False).mean()
    ema_slow = prices.ewm(span=slow, adjust=False).mean()

    macd = ema_fast - ema_slow
    macd_signal = macd.ewm(span=signal, adjust=False).mean()
    macd_hist = macd - macd_signal

    return {
        'macd': macd.iloc[-1],
        'signal': macd_signal.iloc[-1],
        'histogram': macd_hist.iloc[-1]
    }


def calculate_bollinger_bands(prices: pd.Series, period: int = 20, num_std: int = 2) -> Dict[str, float]:
    """Calculate Bollinger Bands"""
    sma = prices.rolling(window=period).mean()
    std = prices.rolling(window=period).std()

    upper = sma + (std * num_std)
    lower = sma - (std * num_std)

    current_price = prices.iloc[-1]
    upper_val = upper.iloc[-1]
    lower_val = lower.iloc[-1]
    middle_val = sma.iloc[-1]

    # Position within bands (0 = at lower, 1 = at upper)
    position = (current_price - lower_val) / (upper_val - lower_val) if (upper_val - lower_val) > 0 else 0.5

    return {
        'upper': upper_val,
        'middle': middle_val,
        'lower': lower_val,
        'position': position
    }


def calculate_stochastic(df: pd.DataFrame, period: int = 14) -> Dict[str, float]:
    """Calculate Stochastic Oscillator"""
    low_min = df['low'].rolling(window=period).min()
    high_max = df['high'].rolling(window=period).max()

    k = 100 * (df['close'] - low_min) / (high_max - low_min)
    d = k.rolling(window=3).mean()

    return {
        'k': k.iloc[-1] if not pd.isna(k.iloc[-1]) else 50.0,
        'd': d.iloc[-1] if not pd.isna(d.iloc[-1]) else 50.0
    }


def calculate_atr(df: pd.DataFrame, period: int = 14) -> float:
    """Calculate Average True Range"""
    high_low = df['high'] - df['low']
    high_close = np.abs(df['high'] - df['close'].shift())
    low_close = np.abs(df['low'] - df['close'].shift())

    tr = pd.concat([high_low, high_close, low_close], axis=1).max(axis=1)
    atr = tr.rolling(window=period).mean()

    return atr.iloc[-1] if not pd.isna(atr.iloc[-1]) else 0.0


def calculate_obv(df: pd.DataFrame) -> float:
    """Calculate On-Balance Volume"""
    obv = (np.sign(df['close'].diff()) * df['volume']).fillna(0).cumsum()
    return obv.iloc[-1]


def calculate_vwap(df: pd.DataFrame, period: int = 20) -> float:
    """Calculate Volume Weighted Average Price"""
    typical_price = (df['high'] + df['low'] + df['close']) / 3
    vwap = (typical_price * df['volume']).rolling(window=period).sum() / df['volume'].rolling(window=period).sum()
    return vwap.iloc[-1] if not pd.isna(vwap.iloc[-1]) else df['close'].iloc[-1]


def extract_features(df: pd.DataFrame, index: int = -1, spy_df: pd.DataFrame = None) -> Dict[str, float]:
    """
    Extract high-signal features for ML prediction
    Focus on features proven to work in quantitative trading
    """
    if len(df) < 200:
        raise ValueError(f"Need at least 200 data points, got {len(df)}")

    # Use all data up to index
    if index == -1:
        data = df.copy()
    else:
        data = df.iloc[:index + 1].copy()

    current = data.iloc[-1]
    close_prices = data['close']
    high_prices = data['high']
    low_prices = data['low']
    volume = data['volume']

    features = {}

    # =================================================================
    # 1. MOMENTUM FEATURES (trend-following)
    # =================================================================

    # Log returns (more stable than simple returns)
    features['log_return_1d'] = np.log(close_prices.iloc[-1] / close_prices.iloc[-2]) if len(close_prices) > 1 else 0
    features['log_return_5d'] = np.log(close_prices.iloc[-1] / close_prices.iloc[-6]) if len(close_prices) > 5 else 0
    features['log_return_21d'] = np.log(close_prices.iloc[-1] / close_prices.iloc[-22]) if len(close_prices) > 21 else 0

    # MACD (momentum indicator)
    macd = calculate_macd(close_prices)
    features['macd_line'] = macd['macd']
    features['macd_signal'] = macd['signal']
    features['macd_histogram'] = macd['histogram']

    # =================================================================
    # 2. MEAN REVERSION FEATURES
    # =================================================================

    # RSI
    features['rsi'] = calculate_rsi(close_prices, 14)
    features['rsi_oversold'] = 1 if features['rsi'] < 30 else 0
    features['rsi_overbought'] = 1 if features['rsi'] > 70 else 0

    # Bollinger Band Z-score (how many std devs from mean)
    bb = calculate_bollinger_bands(close_prices)
    bb_width = bb['upper'] - bb['lower']
    if bb_width > 0:
        features['bb_zscore'] = (current['close'] - bb['middle']) / (bb_width / 4)  # /4 because width is 4 std devs
    else:
        features['bb_zscore'] = 0

    # Price-to-EMA ratios (mean reversion signals)
    ema_20 = close_prices.ewm(span=20, adjust=False).mean().iloc[-1]
    ema_50 = close_prices.ewm(span=50, adjust=False).mean().iloc[-1]
    features['price_to_ema20'] = (current['close'] / ema_20 - 1) if ema_20 > 0 else 0
    features['price_to_ema50'] = (current['close'] / ema_50 - 1) if ema_50 > 0 else 0

    # =================================================================
    # 3. VOLATILITY FEATURES
    # =================================================================

    # Realized volatility (annualized)
    returns = close_prices.pct_change()
    features['realized_vol_5d'] = returns.iloc[-5:].std() * np.sqrt(252) if len(returns) >= 5 else 0
    features['realized_vol_21d'] = returns.iloc[-21:].std() * np.sqrt(252) if len(returns) >= 21 else 0

    # ATR (normalized by price)
    atr = calculate_atr(data, 14)
    features['atr_normalized'] = atr / current['close'] if current['close'] > 0 else 0

    # High-Low range (normalized)
    features['hl_range_normalized'] = (current['high'] - current['low']) / current['close'] if current['close'] > 0 else 0

    # Parkinson volatility estimator (more efficient than close-to-close)
    if len(data) >= 20:
        hl_ratio = np.log(high_prices.iloc[-20:] / low_prices.iloc[-20:])
        parkinson_vol = np.sqrt(np.mean(hl_ratio ** 2) / (4 * np.log(2))) * np.sqrt(252)
        features['parkinson_vol'] = parkinson_vol
    else:
        features['parkinson_vol'] = 0

    # Volatility regime (high vs low)
    vol_21d = features['realized_vol_21d']
    vol_63d = returns.iloc[-63:].std() * np.sqrt(252) if len(returns) >= 63 else vol_21d
    features['vol_regime'] = (vol_21d / vol_63d - 1) if vol_63d > 0 else 0

    # =================================================================
    # 4. LIQUIDITY / MICROSTRUCTURE FEATURES
    # =================================================================

    # Volume z-score (how unusual is today's volume)
    volume_mean = volume.iloc[-20:].mean()
    volume_std = volume.iloc[-20:].std()
    if volume_std > 0:
        features['volume_zscore'] = (current['volume'] - volume_mean) / volume_std
    else:
        features['volume_zscore'] = 0

    # Volume change
    if len(volume) > 1:
        features['volume_change'] = (current['volume'] / volume.iloc[-2] - 1) if volume.iloc[-2] > 0 else 0
    else:
        features['volume_change'] = 0

    # Turnover (volume / price)
    features['turnover'] = current['volume'] * current['close']

    # VWAP deviation (how far from fair value)
    vwap = calculate_vwap(data, 20)
    features['vwap_deviation'] = (current['close'] / vwap - 1) if vwap > 0 else 0

    # Volume trend
    vol_sma_5 = volume.iloc[-5:].mean()
    vol_sma_20 = volume.iloc[-20:].mean()
    features['volume_trend'] = (vol_sma_5 / vol_sma_20 - 1) if vol_sma_20 > 0 else 0

    # =================================================================
    # 5. RELATIVE STRENGTH FEATURES (vs market)
    # =================================================================

    # If SPY data is provided, calculate beta-adjusted returns
    if spy_df is not None and len(spy_df) >= len(data):
        try:
            # Align data
            spy_returns = spy_df['close'].pct_change()
            stock_returns = close_prices.pct_change()

            # Calculate rolling beta (21-day)
            if len(stock_returns) >= 21:
                # Simple linear regression beta
                valid_idx = ~(stock_returns.isna() | spy_returns[-len(stock_returns):].isna())
                if valid_idx.sum() >= 21:
                    recent_stock = stock_returns[valid_idx].iloc[-21:]
                    recent_spy = spy_returns[-len(stock_returns):][valid_idx].iloc[-21:]

                    covariance = recent_stock.cov(recent_spy)
                    spy_variance = recent_spy.var()
                    beta = covariance / spy_variance if spy_variance > 0 else 1.0

                    # Market residual return (alpha)
                    spy_return_1d = spy_returns.iloc[-1] if len(spy_returns) > 0 else 0
                    stock_return_1d = stock_returns.iloc[-1] if len(stock_returns) > 0 else 0
                    features['market_residual_return'] = stock_return_1d - beta * spy_return_1d
                    features['beta_to_spy'] = beta
                else:
                    features['market_residual_return'] = 0
                    features['beta_to_spy'] = 1.0
            else:
                features['market_residual_return'] = 0
                features['beta_to_spy'] = 1.0

            # SPY relative strength
            spy_return_21d = spy_df['close'].iloc[-1] / spy_df['close'].iloc[-22] - 1 if len(spy_df) >= 22 else 0
            stock_return_21d = close_prices.iloc[-1] / close_prices.iloc[-22] - 1 if len(close_prices) >= 22 else 0
            features['relative_strength_vs_spy'] = stock_return_21d - spy_return_21d

        except Exception:
            features['market_residual_return'] = 0
            features['beta_to_spy'] = 1.0
            features['relative_strength_vs_spy'] = 0
    else:
        features['market_residual_return'] = 0
        features['beta_to_spy'] = 1.0
        features['relative_strength_vs_spy'] = 0

    # =================================================================
    # 6. ADDITIONAL CLEAN SIGNALS
    # =================================================================

    # Gap (overnight return)
    if len(data) > 1:
        features['gap'] = (current['open'] / data.iloc[-2]['close'] - 1) if data.iloc[-2]['close'] > 0 else 0
    else:
        features['gap'] = 0

    # Intraday return
    features['intraday_return'] = (current['close'] / current['open'] - 1) if current['open'] > 0 else 0

    # Price action (close relative to high-low range)
    daily_range = current['high'] - current['low']
    if daily_range > 0:
        features['price_position'] = (current['close'] - current['low']) / daily_range
    else:
        features['price_position'] = 0.5

    # Replace NaN and inf values with appropriate defaults
    for key in features:
        if pd.isna(features[key]) or np.isinf(features[key]):
            features[key] = 0.0

    return features


def engineer_dataset(df: pd.DataFrame, look_forward: int = 1, spy_df: pd.DataFrame = None) -> Tuple[pd.DataFrame, pd.Series, List[str], pd.Series]:
    """
    Engineer complete dataset for ML training

    Args:
        df: Stock price data
        look_forward: Days ahead to predict
        spy_df: SPY (market) data for calculating beta and residual returns

    Returns:
        features_df: DataFrame with all features
        targets: Series with target RETURNS (not prices!)
        dates: List of dates
        prices: Series with current prices (for converting returns back to prices)
    """
    features_list = []
    returns_targets = []
    dates = []
    current_prices = []

    # Start from index 200 (need history for indicators)
    for i in range(200, len(df) - look_forward):
        try:
            # Extract SPY data up to same point for fair comparison
            spy_slice = None
            if spy_df is not None:
                spy_slice = spy_df.iloc[:i + 1] if i < len(spy_df) else spy_df

            feature_dict = extract_features(df, i, spy_df=spy_slice)
            features_list.append(feature_dict)

            # Target: RETURN (not absolute price!)
            current_price = df.iloc[i]['close']
            future_price = df.iloc[i + look_forward]['close']
            return_target = (future_price - current_price) / current_price

            returns_targets.append(return_target)
            current_prices.append(current_price)
            dates.append(df.iloc[i]['date'] if 'date' in df.columns else str(df.index[i]))
        except Exception as e:
            print(f"Error extracting features at index {i}: {e}")
            continue

    features_df = pd.DataFrame(features_list)
    returns_series = pd.Series(returns_targets)
    prices_series = pd.Series(current_prices)

    return features_df, returns_series, dates, prices_series


def normalize_features(features_df: pd.DataFrame) -> Tuple[pd.DataFrame, Dict[str, float], Dict[str, float]]:
    """
    Normalize features using z-score normalization

    Returns:
        normalized_df: Normalized features
        means: Dictionary of feature means
        stds: Dictionary of feature standard deviations
    """
    # First, fill any remaining NaN or inf values
    features_df = features_df.replace([np.inf, -np.inf], np.nan)
    features_df = features_df.fillna(0)

    means = features_df.mean().to_dict()
    stds = features_df.std().to_dict()

    # Replace 0 std with 1 to avoid division by zero
    for key in stds:
        if stds[key] == 0 or pd.isna(stds[key]):
            stds[key] = 1.0
        if pd.isna(means[key]):
            means[key] = 0.0

    normalized_df = (features_df - features_df.mean()) / features_df.std().replace(0, 1)

    # Final safety check
    normalized_df = normalized_df.replace([np.inf, -np.inf], 0)
    normalized_df = normalized_df.fillna(0)

    return normalized_df, means, stds

File: lib/ml/feature-engineering.ts
/**
 * Professional ML Feature Engineering
 * Extracts meaningful features from historical stock/ETF data for prediction models
 */

import { PriceData, calculateTechnicalAnalysis } from '../technical-indicators'

export interface MLFeatures {
  // Price features
  closePrice: number
  openPrice: number
  highPrice: number
  lowPrice: number
  volume: number

  // Returns
  returns: number
  logReturns: number
  returns5d: number
  returns10d: number
  returns20d: number

  // Technical indicators
  rsi: number
  macd: number
  macdSignal: number
  macdHist: number
  bollingerUpper: number
  bollingerMiddle: number
  bollingerLower: number
  bollingerPosition: number // Where price is relative to bands

  // Moving averages
  sma5: number
  sma10: number
  sma20: number
  sma50: number
  sma200: number
  ema12: number
  ema26: number

  // Momentum indicators
  roc: number // Rate of change
  momentum: number
  stochK: number
  stochD: number

  // Volatility
  atr: number // Average True Range
  volatility20d: number
  volatility50d: number

  // Volume indicators
  volumeRatio: number // Current volume / avg volume
  obv: number // On-Balance Volume
  vwap: number // Volume Weighted Average Price

  // Pattern features
  priceRange: number
  bodySize: number
  upperShadow: number
  lowerShadow: number
  gapSize: number

  // Trend features
  trend5d: number
  trend10d: number
  trend20d: number
  macdTrend: number
}

export interface EngineereedDataset {
  features: MLFeatures[]
  targets: number[] // Next day's close price
  dates: string[]
  rawPrices: PriceData[]
}

/**
 * Calculate Simple Moving Average
 */
function calculateSMA(prices: number[], period: number): number {
  if (prices.length < period) return prices[prices.length - 1]
  const slice = prices.slice(-period)
  return slice.reduce((sum, p) => sum + p, 0) / period
}

/**
 * Calculate Exponential Moving Average
 */
function calculateEMA(prices: number[], period: number): number {
  if (prices.length < period) return prices[prices.length - 1]

  const multiplier = 2 / (period + 1)
  let ema = calculateSMA(prices.slice(0, period), period)

  for (let i = period; i < prices.length; i++) {
    ema = (prices[i] - ema) * multiplier + ema
  }

  return ema
}

/**
 * Calculate Rate of Change
 */
function calculateROC(prices: number[], period: number): number {
  if (prices.length < period + 1) return 0
  const current = prices[prices.length - 1]
  const past = prices[prices.length - 1 - period]
  return ((current - past) / past) * 100
}

/**
 * Calculate Average True Range (ATR)
 */
function calculateATR(data: PriceData[], period: number): number {
  if (data.length < period) return 0

  const trueRanges: number[] = []
  for (let i = 1; i < data.length; i++) {
    const high = data[i].high
    const low = data[i].low
    const prevClose = data[i - 1].close

    const tr = Math.max(
      high - low,
      Math.abs(high - prevClose),
      Math.abs(low - prevClose)
    )
    trueRanges.push(tr)
  }

  const slice = trueRanges.slice(-period)
  return slice.reduce((sum, tr) => sum + tr, 0) / slice.length
}

/**
 * Calculate On-Balance Volume (OBV)
 */
function calculateOBV(data: PriceData[]): number {
  let obv = 0
  for (let i = 1; i < data.length; i++) {
    if (data[i].close > data[i - 1].close) {
      obv += data[i].volume
    } else if (data[i].close < data[i - 1].close) {
      obv -= data[i].volume
    }
  }
  return obv
}

/**
 * Calculate Volume Weighted Average Price (VWAP)
 */
function calculateVWAP(data: PriceData[], period: number): number {
  if (data.length < period) return data[data.length - 1].close

  const slice = data.slice(-period)
  let sumPriceVolume = 0
  let sumVolume = 0

  for (const bar of slice) {
    const typicalPrice = (bar.high + bar.low + bar.close) / 3
    sumPriceVolume += typicalPrice * bar.volume
    sumVolume += bar.volume
  }

  return sumVolume > 0 ? sumPriceVolume / sumVolume : slice[slice.length - 1].close
}

/**
 * Calculate volatility (standard deviation of returns)
 */
function calculateVolatility(prices: number[], period: number): number {
  if (prices.length < period + 1) return 0

  const returns: number[] = []
  for (let i = prices.length - period; i < prices.length; i++) {
    if (i > 0) {
      returns.push((prices[i] - prices[i - 1]) / prices[i - 1])
    }
  }

  const mean = returns.reduce((sum, r) => sum + r, 0) / returns.length
  const squaredDiffs = returns.map(r => Math.pow(r - mean, 2))
  const variance = squaredDiffs.reduce((sum, d) => sum + d, 0) / returns.length

  return Math.sqrt(variance) * Math.sqrt(252) // Annualized
}

/**
 * Extract features for a single data point
 */
export function extractFeatures(
  data: PriceData[],
  index: number
): MLFeatures | null {
  if (index < 200) return null // Need enough historical data

  const historicalData = data.slice(0, index + 1)
  const current = data[index]
  const closes = historicalData.map(d => d.close)
  const volumes = historicalData.map(d => d.volume)
  const highs = historicalData.map(d => d.high)
  const lows = historicalData.map(d => d.low)

  // Calculate technical analysis
  const technical = calculateTechnicalAnalysis(historicalData)

  // Price features
  const closePrice = current.close
  const openPrice = current.open
  const highPrice = current.high
  const lowPrice = current.low
  const volume = current.volume

  // Returns
  const returns = index > 0 ? (closes[index] - closes[index - 1]) / closes[index - 1] : 0
  const logReturns = index > 0 ? Math.log(closes[index] / closes[index - 1]) : 0
  const returns5d = index >= 5 ? (closes[index] - closes[index - 5]) / closes[index - 5] : 0
  const returns10d = index >= 10 ? (closes[index] - closes[index - 10]) / closes[index - 10] : 0
  const returns20d = index >= 20 ? (closes[index] - closes[index - 20]) / closes[index - 20] : 0

  // Moving averages
  const sma5 = calculateSMA(closes, 5)
  const sma10 = calculateSMA(closes, 10)
  const sma20 = calculateSMA(closes, 20)
  const sma50 = calculateSMA(closes, 50)
  const sma200 = calculateSMA(closes, 200)
  const ema12 = calculateEMA(closes, 12)
  const ema26 = calculateEMA(closes, 26)

  // Momentum
  const roc = calculateROC(closes, 10)
  const momentum = index >= 10 ? closes[index] - closes[index - 10] : 0

  // Volatility
  const atr = calculateATR(historicalData, 14)
  const volatility20d = calculateVolatility(closes, 20)
  const volatility50d = calculateVolatility(closes, 50)

  // Volume
  const avgVolume = calculateSMA(volumes, 20)
  const volumeRatio = avgVolume > 0 ? volume / avgVolume : 1
  const obv = calculateOBV(historicalData)
  const vwap = calculateVWAP(historicalData, 20)

  // Pattern features
  const priceRange = highPrice - lowPrice
  const bodySize = Math.abs(closePrice - openPrice)
  const upperShadow = highPrice - Math.max(openPrice, closePrice)
  const lowerShadow = Math.min(openPrice, closePrice) - lowPrice
  const gapSize = index > 0 ? openPrice - data[index - 1].close : 0

  // Trend features (slope of linear regression)
  const trend5d = calculateTrend(closes.slice(-5))
  const trend10d = calculateTrend(closes.slice(-10))
  const trend20d = calculateTrend(closes.slice(-20))
  const macdTrend = technical.macd.histogram > 0 ? 1 : -1

  // Bollinger Band position
  const bollingerRange = technical.bollingerBands.upper - technical.bollingerBands.lower
  const bollingerPosition = bollingerRange > 0
    ? (closePrice - technical.bollingerBands.lower) / bollingerRange
    : 0.5

  return {
    closePrice,
    openPrice,
    highPrice,
    lowPrice,
    volume,
    returns,
    logReturns,
    returns5d,
    returns10d,
    returns20d,
    rsi: technical.rsi.value,
    macd: technical.macd.macd,
    macdSignal: technical.macd.signal,
    macdHist: technical.macd.histogram,
    bollingerUpper: technical.bollingerBands.upper,
    bollingerMiddle: technical.bollingerBands.middle,
    bollingerLower: technical.bollingerBands.lower,
    bollingerPosition,
    sma5,
    sma10,
    sma20,
    sma50,
    sma200,
    ema12,
    ema26,
    roc,
    momentum,
    stochK: technical.stochastic.k,
    stochD: technical.stochastic.d,
    atr,
    volatility20d,
    volatility50d,
    volumeRatio,
    obv,
    vwap,
    priceRange,
    bodySize,
    upperShadow,
    lowerShadow,
    gapSize,
    trend5d,
    trend10d,
    trend20d,
    macdTrend,
  }
}

/**
 * Calculate trend (slope) using linear regression
 */
function calculateTrend(prices: number[]): number {
  if (prices.length < 2) return 0

  const n = prices.length
  let sumX = 0
  let sumY = 0
  let sumXY = 0
  let sumX2 = 0

  for (let i = 0; i < n; i++) {
    sumX += i
    sumY += prices[i]
    sumXY += i * prices[i]
    sumX2 += i * i
  }

  const slope = (n * sumXY - sumX * sumY) / (n * sumX2 - sumX * sumX)
  return slope
}

/**
 * Engineer complete dataset for ML training
 */
export function engineerDataset(
  historicalData: PriceData[],
  lookForward: number = 1
): EngineereedDataset {
  const features: MLFeatures[] = []
  const targets: number[] = []
  const dates: string[] = []

  for (let i = 200; i < historicalData.length - lookForward; i++) {
    const feature = extractFeatures(historicalData, i)
    if (feature) {
      features.push(feature)
      targets.push(historicalData[i + lookForward].close) // Next day's close
      dates.push(historicalData[i].date)
    }
  }

  return {
    features,
    targets,
    dates,
    rawPrices: historicalData,
  }
}

/**
 * Normalize features for neural networks
 */
export function normalizeFeatures(features: MLFeatures[]): {
  normalized: number[][]
  means: number[]
  stds: number[]
  featureNames: string[]
} {
  const featureNames = Object.keys(features[0]) as (keyof MLFeatures)[]
  const numFeatures = featureNames.length

  // Calculate means and standard deviations
  const means: number[] = new Array(numFeatures).fill(0)
  const stds: number[] = new Array(numFeatures).fill(0)

  // Calculate means
  for (const feature of features) {
    featureNames.forEach((name, idx) => {
      means[idx] += feature[name]
    })
  }
  means.forEach((_, idx) => {
    means[idx] /= features.length
  })

  // Calculate standard deviations
  for (const feature of features) {
    featureNames.forEach((name, idx) => {
      stds[idx] += Math.pow(feature[name] - means[idx], 2)
    })
  }
  stds.forEach((_, idx) => {
    stds[idx] = Math.sqrt(stds[idx] / features.length)
    if (stds[idx] === 0) stds[idx] = 1 // Prevent division by zero
  })

  // Normalize
  const normalized: number[][] = features.map(feature => {
    return featureNames.map((name, idx) => {
      return (feature[name] - means[idx]) / stds[idx]
    })
  })

  return { normalized, means, stds, featureNames }
}

/**
 * Denormalize predictions
 */
export function denormalizePredictions(
  normalized: number[],
  mean: number,
  std: number
): number[] {
  return normalized.map(val => val * std + mean)
}
